### PR TITLE
ipn/ipnlocal: add c2n /debug/pprof/allocs endpoint

### PR DESCRIPTION
**Pull request by oxtoacart**
This allows C2N clients to obtain sampled information about memory allocations since the program started.

We now also force GC on logheap.

Updates tailscale/corp#18514

### CODE REVIEW

File: ipn/ipnlocal/c2n.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ var c2nHandlers = map[methodAndPath]c2nHandler{
 	req("/debug/metrics"):           handleC2NDebugMetrics,
 	req("/debug/component-logging"): handleC2NDebugComponentLogging,
 	req("/debug/logheap"):           handleC2NDebugLogHeap,
+	req("/debug/logallocs"):         handleC2NDebugLogAllocs,
```
Comment (bradfitz) [inline ipn/ipnlocal/c2n.go:51]: The "log" here is weird. I imagine you're copying "logheap" but that is weird due to history: it predated c2n so the only way we could get pprof data out of the iOS app was to log it and pick it up later async.

Why don't we go with:

```go
	req("/debug/logheap"):           handleC2NDebugPprofHeap,
	req("/debug/pprof/heap"):         handleC2NDebugPprofHeap,
	req("/debug/pprof/allocs"):         handleC2NDebugPprofAllocs,
```

(Moving to /pprof/ but keeping the legacy path too)

File: ipn/ipnlocal/c2n.go [hunk 2]
```diff
@@ -175,9 +176,22 @@ func handleC2NDebugLogHeap(b *LocalBackend, w http.ResponseWriter, r *http.Reque
 		http.Error(w, "not implemented", http.StatusNotImplemented)
 		return
 	}
+	runtime.GC()
```
Comment (bradfitz) [inline ipn/ipnlocal/c2n.go:179]: unrelated to this PR.

also, I'd sent https://github.com/tailscale/tailscale/pull/11479 to make us match the Go option to do this if requested.

Comment (oxtoacart) [reply ipn/ipnlocal/c2n.go]: Removed

File: ipn/ipnlocal/c2n.go (hunk 1 above)
Comment (oxtoacart) [reply ipn/ipnlocal/c2n.go]: Done (parametrically)

File: ipn/ipnlocal/c2n.go [hunk 3]
```diff
@@ -48,6 +48,8 @@ var c2nHandlers = map[methodAndPath]c2nHandler{
 	req("/debug/metrics"):           handleC2NDebugMetrics,
 	req("/debug/component-logging"): handleC2NDebugComponentLogging,
 	req("/debug/logheap"):           handleC2NDebugLogHeap,
```
Comment (bradfitz) [inline ipn/ipnlocal/c2n.go:50]: change this line's value too and delete the `handleC2NDebugLogHeap` variable? Oh, but then you'd be deleting the "gc" parameter which this isn't rebased to include?

In any case, /debug/logheap and /debug/pprof/heap should have identical behavior (and both support the gc parameter. or at least the pprof version should)

**Review by bradfitz — APPROVED** [review]
c2n isn't a package, so commit message: `ipn/ipnlocal: add c2n /debug/pprof/allocs endpoint`

(the part before the colon is typically a directory / Go package)

File: ipn/ipnlocal/c2n.go (hunk 3 above)
Comment (oxtoacart) [reply ipn/ipnlocal/c2n.go]: I decided to go all out here and just make the `/debug/pprof/` endpoint behave exactly like the real `/debug/pprof`. This required me to introduce a ServeMux, but I left the old pattern matching in place for compatibility.

**Review by bradfitz — COMMENTED** [review]
Sorry, but please revert back to the earlier thing with the requested changed. We can't open up all of pprof for security/privacy reasons. We're intentionally chose a curated subset of debugging endpoints that are reasonable to expose to us. That's why for example we scrub memory addresses in the goroutines handlers. This would negate all that, making it wide open.

File: ipn/ipnlocal/c2n_pprof.go [hunk 4]
```diff
@@ -10,6 +10,8 @@ import (
 	"runtime"
 	"runtime/pprof"
 	"strconv"
+
+	hpprof "tailscale.com/tempfork/pprof"
```
Comment (bradfitz) [inline ipn/ipnlocal/c2n_pprof.go:14]: actually I'd like to delete this whole forked package. It was only one caller left inside corp and it's redundant there now that we have c2n. Want to just do it by hand like in your earlier version and then also delete this package?

The net/http/pprof package is super minimal. It's not worth trying to reuse that code just to fight text/template and reflect.MethodByName coming along for the ride.

**Pull request merged**